MediaBrowser.Server.Mono/ApplicationPathHelper.cs

```diff
@@ -1,44 +1,56 @@
 using System.IO;
 using System.Runtime.InteropServices;
+using EmbyServer;
 
 namespace MediaBrowser.Server.Mono
 {
-    public static class ApplicationPathHelper
+    public class ApplicationPathHelper
     {
+        private readonly string _debugProgramDataPath;
+        private readonly string _releaseProgramDataPath;
+        private readonly IOperatingSystemInformationLookup _osInfoLookup;
+        private readonly bool _debug;
+
+        public ApplicationPathHelper(
+            string debugProgramDataPath,
+            string releaseProgramDataPath,
+            IOperatingSystemInformationLookup osInfoLookup,
+            bool debug)
+        {
+            _debugProgramDataPath = debugProgramDataPath;
+            _releaseProgramDataPath = releaseProgramDataPath;
+            _osInfoLookup = osInfoLookup;
+            _debug = debug;
+        }
+       
         /// <summary>
         /// Gets the path to the application's ProgramDataFolder
         /// </summary>
         /// <returns>System.String.</returns>
-        public static string GetProgramDataPath(string applicationPath)
+        public string GetProgramDataPath(string applicationPath)
         {
-            var useDebugPath = false;
+            var programDataPath = _debug ? _debugProgramDataPath : _releaseProgramDataPath;
 
-#if DEBUG
-            useDebugPath = true;
-#endif
-
-            var programDataPath = useDebugPath ? 
-                ConfigurationManager.AppSettings["DebugProgramDataPath"] : 
-                ConfigurationManager.AppSettings["ReleaseProgramDataPath"];
-
-            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-            {
-                programDataPath = programDataPath.Replace("%ApplicationData%", Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData));
-            }
-            else
-            {
-                programDataPath = programDataPath.Replace("%ApplicationData%", "/var/lib");
-            }
+            var isWindows = _osInfoLookup.GetOperatingSystem() == OSPlatform.Windows;
+            var windowsAppDataPath = _osInfoLookup.GetApplicationDataPath();
+            // GetApplicationDataPath() will return $HOME/... on *nix, so hard code an alternative here.
+            const string unixAppDataPath = "/var/lib";
             
+            programDataPath = isWindows
+                ? programDataPath.Replace("%ApplicationData%", windowsAppDataPath)
+                : programDataPath.Replace("%ApplicationData%", unixAppDataPath);
 
             programDataPath = programDataPath
-                .Replace('/', Path.DirectorySeparatorChar)
-                .Replace('\\', Path.DirectorySeparatorChar);
+                // Replace empty directories caused by trailing slashes
+                .Replace("//", "/")
+                .Replace("\\\\", "\\")
+                // Correct directory separator
+                .Replace('/', _osInfoLookup.GetDirectorySeparatorChar())
```

Comment:
I think you need to swap replacing '/' with os separator and removing double slashes, otherwise paths like `c:\foo\/bar` won't get fixed correctly.

Reply:
Good shout, I'll add a test case for this, as well.

Reply:
FWIW this should now be largely unneeded. As of windows 8 (and I think backported to 7 I'd need to check) / and \ are interchangeable to windows now at the api level.

Reply:
I thought / and \ worked on Windows for ages (do not remember if that was the case for 9x, but I think it was for XP). Still "default" separator is \, and I think it's better to stick to what platform considers default even if we need to write a bit more code.